flasc/analysis/expected_power_analysis_utilities.py

```diff
@@ -0,0 +1,484 @@
+        test_cols (List[str]): A list of column names to calculate the covariance of
+        bin_cols_with_df_name (List[str]): A list of column names to bin the dataframes by.
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the covariance matrix.
+    """
+    # Generate all pairs of columns (including same column pairs)
+    col_pairs = list(product(test_cols, test_cols))
+
+    # Create expressions for all covariances
+    cov_exprs = [pl.cov(pair[0], pair[1]).alias(f"cov_{pair[0]}_{pair[1]}") for pair in col_pairs]
+
+    # Compute covariances
+    df_cov = df_.group_by(bin_cols_with_df_name).agg(cov_exprs)
+
+    # Get the number of points for each pair
+    df_n = _get_num_points_pair(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    # Join the number of points to the covariance matrix
+    df_cov = df_cov.join(df_n, on=bin_cols_with_df_name, how="left")
+
+    # Enforce that each ws/wd bin combination has to appear in all dataframes
+    bin_cols_without_df_name = [c for c in bin_cols_with_df_name if c != "df_name"]
+    num_df = df_["df_name"].n_unique()
+    df_cov = df_cov.filter(pl.len().over(bin_cols_without_df_name) == num_df)
+
+    return df_cov
+
+
+def _null_and_sync_covariance(
+    df_cov: pl.DataFrame,
+    test_cols: List[str],
+    uplift_pairs: List[List[str]],
+    bin_cols_without_df_name: List[str] = ["wd_bin", "ws_bin"],
+) -> pl.DataFrame:
+    """Apply null values to the covariance matrix and synchronize across uplift pairs.
+
+    Args:
+        df_cov (pl.DataFrame): A polars dataframe with the covariance matrix
+        test_cols (List[str]): A list of column names to calculate the covariance of
+        uplift_pairs (List[List[str]]): A list of the df_name values to copy the nans from
+        bin_cols_without_df_name (List[str]): A list of column names to bin the dataframes by.
+            Defaults to ["wd_bin", "ws_bin"].
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the null values applied and synchronized across
+            uplift pairs.
+    """
+    # Get the names of all the covariance and num_points columns
+    cov_cols = [f"cov_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+    n_cols = [f"count_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+
+    # Convert all NaN to None
+    df_cov = df_cov.fill_nan(None)
+
+    # Loop over cov_cols and n_cols
+    for cov_col, n_col in zip(cov_cols, n_cols):
+        # Wherever n_col 0, 1  set the corresponding cov_col to NaN
+        df_cov = df_cov.with_columns(
+            pl.when(pl.col(n_col) < 2)  # | pl.col(n_col).is_null())
+            .then(None)
+            .otherwise(pl.col(cov_col))
+            .alias(cov_col)
+        )
+
+        # Wherever n_col is null, set cov_col to null
+        df_cov = df_cov.with_columns(
+            pl.when(pl.col(n_col).is_null()).then(None).otherwise(pl.col(cov_col)).alias(cov_col)
+        )
```

Comment:
It seems like some of the null setting in this function is duplicating what is already done in _fill_cov_with_var. We should check to see what is necessary.

Reply:
Possibly, I was thinking if things get copied over could be good to double check in that it doesn't hurt at least

Reply:
Instead of setting cov to Null when n_col <2 or n_col is Null, which should already be the case, we can add a test to make sure if number of overlapping points is < 2, the computed covariance will be Null for that turbine pair.